torchmetrics/image/map.py

```diff
@@ -0,0 +1,369 @@
+    (xmin-top left, ymin-top left, xmax-bottom right, ymax-bottom right).
+    See the `update` function for more information about the input format to this metric.
+
+    .. note::
+        This metric is a wrapper for the
+        `pycocotools <https://github.com/cocodataset/cocoapi/tree/master/PythonAPI/pycocotools>`_,
+        which is a standard implementation for the mAP metric for object detection. Using this metric
+        therefore requires you to have `pycocotools` installed. Please install with `pip install pycocotools` or
+        `pip install torchmetrics[image]`
+
+    .. note::
+        As the pycocotools library cannot deal with tensors directly, all results have to be transfered
+        to the CPU, this might have an performance impact on your training
+
+    Args:
+        num_classes:
+            Number of classes, required for mAP values per class. default: 0 (deactivate)
+        compute_on_step:
+            Forward only calls ``update()`` and return None if this is set to False. default: False
+        dist_sync_on_step:
+            Synchronize metric state across processes at each ``forward()``
+            before returning the value at the step. default: False
+        process_group:
+            Specify the process group on which synchronization is called. default: None (which selects the entire world)
+
+    Raises:
+        ValueError:
+            If ``pycocotools`` is not installed
+        ValueError:
+            If ``num_classes`` is not an integer larger or equal to 0
+    """
+
+    def __init__(
+        self,
+        num_classes: int = 0,
+        compute_on_step: bool = False,
+        dist_sync_on_step: bool = False,
+        process_group: Optional[Any] = None,
+    ) -> None:
+        super().__init__(
+            compute_on_step=compute_on_step,
+            dist_sync_on_step=dist_sync_on_step,
+            process_group=process_group,
+        )
+        if not _PYCOCOTOOLS_AVAILABLE:
+            raise ValueError(
+                "`MAP` metric requires that `pycocotools` installed. Please install"
+                " with `pip install pycocotools` or `pip install torchmetrics[image]`"
+            )
+
+        self.add_state("average_precision", default=torch.tensor(data=[], dtype=torch.float), dist_reduce_fx="mean")
+        self.add_state("average_recall", default=torch.tensor(data=[], dtype=torch.float), dist_reduce_fx="mean")
+        if not (isinstance(num_classes, int) and num_classes >= 0):
+            raise ValueError("Expected argument `num_classes` to be a integer larger or equal to 0")
+        self.num_classes = num_classes
+
+        self.add_state("detection_boxes", default=[])
+        self.add_state("detection_scores", default=[])
+        self.add_state("detection_labels", default=[])
+        self.add_state("groundtruth_boxes", default=[])
+        self.add_state("groundtruth_labels", default=[])
+
+        for class_id in range(num_classes):
```

Comment:
If I'm not mistaken, a state variable can also be a list. In that case you could create two state variables, ``ap_per_class`` and ``ar_per_class``, with ``num_classes`` entries, and you don't have to use ``getattr()`` / ``setattr()``.

Reply:
Agreed, that would be much nicer. But what would be the correct 'dist_reduce_fx' for this?
In `DDP` mode we would have to calculate the mean over each value.

Reply:
Maybe this problem related to the discussion in https://github.com/PyTorchLightning/metrics/pull/467#discussion_r714596707

Reply:
Resolved as of https://github.com/PyTorchLightning/metrics/pull/467#discussion_r714596707